src/internal/ans.ts

```diff
@@ -0,0 +1,177 @@
+  "and can only contain lowercase a-z, 0-9, and hyphens.",
+  "A name may not start or end with a hyphen.",
+].join(" ");
+
+/**
+ *
+ * @param fragment A fragment of a name, either the domain or subdomain
+ * @returns boolean indicating if the fragment is a valid fragment
+ */
+export function isValidANSSegment(fragment: string): boolean {
+  if (!fragment) return false;
+  if (fragment.length < 3) return false;
+  if (fragment.length > 63) return false;
+  // only lowercase a-z and 0-9 are allowed, along with -. a domain may not start or end with a hyphen
+  if (!/^[a-z\d][a-z\d-]{1,61}[a-z\d]$/.test(fragment)) return false;
+  return true;
+}
+
+/**
+ * Checks if an ANS name is valid or not
+ *
+ * @param name A string of the domain name, can include or exclude the .apt suffix
+ */
+export function isValidANSName(name: string): { domainName: string; subdomainName?: string } {
+  const [first, second, ...rest] = name.replace(/\.apt$/, "").split(".");
+
+  if (rest.length > 0) {
+    throw new Error(`${name} is invalid. A name can only have two parts, a domain and a subdomain separated by a "."`);
+  }
+
+  if (!isValidANSSegment(first)) {
+    throw new Error(`${first} is not valid. ${VALIDATION_RULES_DESCRIPTION}`);
+  }
+
+  if (second && !isValidANSSegment(second)) {
+    throw new Error(`${second} is not valid. ${VALIDATION_RULES_DESCRIPTION}`);
+  }
+
+  return {
+    domainName: second || first,
+    subdomainName: second ? first : undefined,
+  };
+}
+
+export const LOCAL_ANS_ACCOUNT_PK =
+  (process.env.ANS_TEST_ACCOUNT_PRIVATE_KEY as string) ||
+  "0x37368b46ce665362562c6d1d4ec01a08c8644c488690df5a17e13ba163e20221";
+export const LOCAL_ANS_ACCOUNT_ADDRESS =
+  (process.env.ANS_TEST_ACCOUNT_ADDRESS as string) ||
+  "0x585fc9f0f0c54183b039ffc770ca282ebd87307916c215a3e692f2f8e4305e82";
+
+const NetworkToAnsContract: Record<Network, string | null> = {
+  [Network.TESTNET]: "0x5f8fd2347449685cf41d4db97926ec3a096eaf381332be4f1318ad4d16a8497c",
+  [Network.MAINNET]: "0x867ed1f6bf916171b1de3ee92849b8978b7d1b9e0a8cc982a3d19d535dfd9c0c",
+  [Network.LOCAL]: LOCAL_ANS_ACCOUNT_ADDRESS,
+  [Network.CUSTOM]: null,
+  [Network.DEVNET]: null,
+};
+
+function getRouterAddress(aptosConfig: AptosConfig): string {
+  const address = NetworkToAnsContract[aptosConfig.network];
+  if (!address) throw new Error(`The ANS contract is not deployed to ${aptosConfig.network}`);
+  return address;
+}
+
+const Some = <T>(value: T): MoveValue => ({ vec: [value] } as any);
+const None = (): MoveValue => ({ vec: [] } as any);
+// != here is intentional, we want to check for null and undefined
+// eslint-disable-next-line eqeqeq
+const Option = <T>(value: T | undefined | null): MoveValue => (value != undefined ? Some(value) : None());
+
+const unwrapOption = <T>(option: any): T | undefined => {
+  if (!!option && typeof option === "object" && "vec" in option && Array.isArray(option.vec)) {
+    return option.vec[0];
+  }
+
+  return undefined;
+};
```

Comment:
We'll have to come back and revisit this for view functions as a whole.  I'm hoping this PR https://github.com/aptos-labs/aptos-core/pull/10761 will allow for arguments to be parsed the same as entry functions